CCPA/Data Deletion Request Handling.md

```diff
@@ -0,0 +1,328 @@
+         i.style.cssText = 'display:none';
+         i.name = '__uspapiLocator';
+         document.body.appendChild(i);
+        }
+        else
+        {
+         window.setTimeout('uspapi_addFrame()', 10);
+        }
+       }
+      }
+      
+      function uspapi_stub()
+      {
+       var b = arguments;
+       __uspapi.a = __uspapi.a || [];
+       if(!b.length)
+       {
+        return __uspapi.a;
+       }
+       else
+       {
+        __uspapi.a.push([].slice.apply(b));
+       }
+      }
+      
+      function uspapi_msghandler(event)
+      {
+       var msgIsString = typeof event.data === 'string';
+       try
+       {
+        var json = msgIsString ? JSON.parse(event.data) : event.data;
+       }
+       catch(e)
+       {
+        var json = null;
+       }
+       if(typeof (json) === 'object' && json !== null && '__uspapiCall' in json)
+       {
+        var i = json.__uspapiCall;
+        window.__uspapi(i.command, i.version, function (retValue, success) 
+        {
+         var returnMsg = {
+         '__uspapiReturn': {
+          'returnValue': retValue,
+          'success'    : success,
+          'callId'     : i.callId
+          }
+         };
+         event.source.postMessage(msgIsString ? JSON.stringify(returnMsg) : returnMsg, '*');
+        }, i.parameter);
+       }
+      }
+      
+      uspapi_addFrame();
+      if(!('__uspapi' in window) || 
+         (typeof (window['__uspapi']) !== 'function' && 
+          typeof (window['__uspapi']) !== 'object' && 
+          (typeof (window['__uspapi']) === 'undefined' || 
+           window['__uspapi'] !== null)))
+      {
+       window['__uspapi'] = uspapi_stub;
+       window['__uspapi'].msgHandler = uspapi_msghandler;
+       if(window.addEventListener)
+       {
+        window.addEventListener('message', uspapi_msghandler, false);
+       }
+       else
+       {
+        window.attachEvent('onmessage', uspapi_msghandler);
+       }
+      }
+    </script>
+    <!-- include the vendor deletion scripts into the page: -->
+    <script src="https://vendorx.com/ccpa-delete-function.js"></script>
+  </head>
+  <body>
+    <button id="ccpa-delete">Delete my personal data</button>
+    <script>
+      document.getElementById('ccpa-delete').addEventListener('click', function ()
+      {
+       __uspapi('performDeletion', 1, null);
+      }, false);
+    </script>
+  </body>
+</html>
+```
+
+Vendor script (at `https://vendorx.com/ccpa-delete-function.js`)
+
+```
+// find the __uspapi frame
+var f = window;
+var cmpFrame;
+var cmpCallbacks = {};
+while (!cmpFrame) {
+  try {
+    if (f.frames['__uspapiLocator']) {
+      cmpFrame = f;
+    }
+  } catch (e) {}
+  if (f === window.top) {
+    break;
+  }
+  f = f.parent;
+}
+
+/* Set up a __uspapi function to do the postMessage and
+stash the callback.
+This function behaves (from the caller's perspective)
+identically to the in-frame __uspapi call */
+window.__uspapi = function(cmd, ver, callback, param) {
```

Comment:
This has always somewhat confused me, mainly because I'm not a JS expert (more of a novice). Does this not overwrite the actual `__uspapi` function with this one?

Reply:
yes it does. but it should only be called/executed if a __usapi function is NOT found in the page. it then serves as a proxy for communicating between iframes

Reply:
not sure the brackets are set correct in the example